### PR TITLE
Wip/alphavantage

### PR DESCRIPTION
**Pull request by james-zhuang**
Uploading financial assistant project

### CODE REVIEW

File: staging/proj.finassistant/manifest.tt [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  query create_report(in req data: Object
+    		   #_[prompt="What do you want to name your report?"]
+               )
+  #_[canonical="Create report"]
+  #_[confirmation="Created report ${name}"]
+  #[doc="create report"];
+
+  query display_report(in req name: String
+    		   #_[prompt="What report do you want?"]
+               )
```
Comment (gcampax) [inline staging/proj.finassistant/manifest.tt:133]: This function needs to return "something".

I am not sure exactly how to implement this feature. One way is to return a link to the report that your frontend knows how to display. Another way is to return a ThingTalk program with the report, so you reexecute the ThingTalk program in your frontend and get the updated data.
If you go with the link/thingtalk program, you'd write something like:
```
entity report;
list query report(out id : Entity(proj.financeassistant:report),
                          out created_on : Date,
                          out companies : Array(Entity(tt:stock_id)),
                          out fields : Array(Enum(revenue,earnings,market_cap, ...))
                          out link : Entity(tt:url))
```
Note that everything is an output. So you can pull up the report by name with:
```
@proj.financeassistant.report(), id =~ "my report";
```
(`id` is of type Entity so it has both an identifier and a name, `=~` is the soft-match operator and looks at the name)
but you can also pull up all reports of a given company:
```
@proj.financeassistant.report(), contains(companies, "goog"^^tt:stock_id);
```
or all market_cap reports:
```
@proj.financeassistant.report(), contains(fields, enum market_cap);
```
These are all relatively easy to express in natural language, and if you write the query this way and write enough annotations on each output parameter, you don't have to do anything for NL, Genie takes care of making up all the possible combinations. Also implementation wise it's quite easy: you can just dump the whole list of reports out of the Thingpedia function, and let the ThingTalk runtime do the actual filtering.

File: staging/proj.finassistant/manifest.tt [hunk 2]
```diff
@@ -0,0 +1,139 @@
+// Copyright 2016-2020 The Board of Trustees of the Leland Stanford Junior University
```
Comment (gcampax) [inline staging/proj.finassistant/manifest.tt:1]: You're not a Stanford employee / RA. Are you donating copyright? Or do you want to retain it?

File: staging/proj.finassistant/manifest.tt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+class @proj.finassistant
+#_[thingpedia_name="Finance Assistant"]
+#_[thingpedia_description="Research assistant leveraging Alpha Vantage APIs to help users quickly find relevant financial data to support their investing decisions"]
+#[subcategory="media"]
+#[license="BSD-3-Clause"]
+#[license_gplcompatible=true]
+#[website="https://www.alphavantage.co"]
+{
+  import loader from @org.thingpedia.v2();
+  import config from @org.thingpedia.config.none(api_key="VPJXG30AV7OQAM22");
```
Comment (gcampax) [inline staging/proj.finassistant/manifest.tt:39]: Don't put the API key publicly like this! Make this `api_key=$?` and create a secrets.json file containing the real API key (`{"api_key": "..."}`). Do not commit that file. Send me the key privately and then I will commit an encrypted copy of that file.

(Amend this commit and force-push to remove the key from git history)

File: staging/proj.finassistant/index.js [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+//
+// Author: James zhuang <james.zhuang@cs.stanford.edu>
+"use strict";
+
+const Tp = require('thingpedia');
+const URL = "https://www.alphavantage.co/query?function=";
+const io = require('socket.io-client');
+const server = 'http://localhost:3000';
+const moment = require('moment');
+const fs = require('fs');
```
Comment (gcampax) [inline staging/proj.finassistant/index.js:40]: Alright, so, right off the bat, I can tell you this architecture will not work. Almond skills can be executed in a variety of environments, most commonly in our server at almond.stanford.edu, and you won't have the ability to open ports or write to the local filesystem (well, you do have that ability, it's just that nothing else can see your local filesystem so it's not very useful).

Any data coming out of the skill needs to go through the Almond API, that is, it needs to be declared in the manifest and returned by the Thingpedia API.

Now, currently the Almond API doesn't expose the raw results unless you use the low-level ThingTalk execution API (which bypasses the natural language). But see https://github.com/stanford-oval/genie-toolkit/pull/420 for a PR that introduces exactly the API you need: you keep using the conversation API (stateful with websocket or stateless with REST and conversation ID), which takes text in and text out, but you also get a message every time Almond executes a ThingTalk program, with the result of that program. You can use that message as clue that the user wants to display a new chart, and display the chart.

You should move the logic for maintaining the dashboard (appending / replacing / clearing) to the server that the browser will connect to.

**Review by gcampax — CHANGES_REQUESTED** [review]
Sorry for the delay in the review. This will need some work to fit with the Almond architecture.

File: staging/proj.finassistant/manifest.tt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+// OF THE POSSIBILITY OF SUCH DAMAGE.
+
+class @proj.finassistant
```
Comment (gcampax) [inline staging/proj.finassistant/manifest.tt:30]: We should choose a better name. Why not `co.alphavantage`? Kill the existing one, which is subsumed by this skill entirely, and clean up. You get to inherit all the existing training and test data for NL as well.

File: staging/proj.finassistant/index.js (hunk 4 above)
Comment (james-zhuang) [reply staging/proj.finassistant/index.js:40]: Yes, will follow this approach and reorg a lot of the existing code

**Comment by gcampax** [comment]
Let's merge this, so we can train a model and keep going.

**Pull request merged**